src/libopensc/card-belpic.c

```diff
@@ -577,6 +577,41 @@ static int str2lang(sc_context_t *ctx, char *lang)
 	return -1;
 }
 
+static int get_applet_version(sc_card_t *card)
+{
+	sc_apdu_t apdu;
+	u8 carddata[28];
+	u8 carddata_cmd[] = { 0x80, 0xE4, 0x00, 0x00, 0x1C };
+	int r;
+
+	r = sc_bytes2apdu(card->ctx, carddata_cmd, sizeof(carddata_cmd), &apdu);
+	if(r) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "bytes to APDU conversion failed: %d\n", r);
+		return r;
+	}
+
+	apdu.resp = carddata;
+	apdu.resplen = sizeof(carddata);
+
+	r = sc_transmit_apdu(card, &apdu);
+	if(r) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "GetCardData command failed: %d\n", r);
+		return r;
+	}
+
+	r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+	if(r) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "GetCardData: card returned %d\n", r);
+		return r;
+	}
+	if(apdu.resplen < sizeof(carddata)) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "GetCardData: card returned %d bytes rather than expected %d\n", apdu.resplen, sizeof(carddata));
+		return SC_ERROR_WRONG_LENGTH;
+	}
+
+	return (int)(carddata[21]);
```

Comment:
On Fri, Mar 27, 2015 at 10:26:05AM -0700, Frank Morgner wrote:
> In src/libopensc/card-belpic.c:
> 
> > +     if(r) {
> > +             sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "GetCardData command failed: %d\n", r);
> > +             return r;
> > +     }
> > +
> > +     r = sc_check_sw(card, apdu.sw1, apdu.sw2);
> > +     if(r) {
> > +             sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "GetCardData: card returned %d\n", r);
> > +             return r;
> > +     }
> > +     if(apdu.resplen < sizeof(carddata)) {
> > +             sc_debug(card->ctx, SC_LOG_DEBUG_NORMAL, "GetCardData: card returned %d bytes rather than expected %d\n", apdu.resplen, sizeof(carddata));
> > +             return SC_ERROR_WRONG_LENGTH;
> > +     }
> > +
> > +     return (int)(carddata[21]);
> 
> I see. Would you mind using a define instead of a magic value to access this
> byte?

Sure. I'll change the code to that effect.

-- 
It is easy to love a country that is famous for chocolate and beer

  -- Barack Obama, speaking in Brussels, Belgium, 2014-03-26